runtime/vm/OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker.cpp

```diff
@@ -0,0 +1,162 @@
+#include "OutOfLineINL.hpp"
+#include "FFITypeHelpers.hpp"
+#include "AtomicSupport.hpp"
+
+extern "C" {
+
+#if JAVA_SPEC_VERSION >= 16
+/* jdk.internal.foreign.abi.ProgrammableInvoker: private static synchronized native void resolveRequiredFields(); */
+VM_BytecodeAction
+OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker_resolveRequiredFields(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	J9JavaVM *vm = currentThread->javaVM;
+	J9ConstantPool *jclConstantPool = (J9ConstantPool *)vm->jclConstantPool;
+	const int cpEntryNum = 2;
+	U_16 cpIndex[cpEntryNum] = {
+				J9VMCONSTANTPOOL_JDKINTERNALFOREIGNABIPROGRAMMABLEINVOKER_CIFNATIVETHUNKADDR,
+				J9VMCONSTANTPOOL_JDKINTERNALFOREIGNABIPROGRAMMABLEINVOKER_ARGTYPESADDR
+			};
+
+	for (int i = 0; i < cpEntryNum; i++) {
+		J9RAMFieldRef *cpFieldRef = ((J9RAMFieldRef*)jclConstantPool) + cpIndex[i];
+		UDATA const flags = cpFieldRef->flags;
+		UDATA const valueOffset = cpFieldRef->valueOffset;
+
+		if (!VM_VMHelpers::instanceFieldRefIsResolved(flags, valueOffset)) {
+			resolveInstanceFieldRef(currentThread, NULL, jclConstantPool, cpIndex[i], J9_RESOLVE_FLAG_NO_THROW_ON_FAIL | J9_RESOLVE_FLAG_JCL_CONSTANT_POOL, NULL);
+			if (VM_VMHelpers::exceptionPending(currentThread)) {
+				rc = GOTO_THROW_CURRENT_EXCEPTION;
+				goto done;
+			}
+		}
+	}
+
+done:
+	VM_OutOfLineINL_Helpers::returnVoid(currentThread, 0);
+	return rc;
+}
+
+/* jdk.internal.foreign.abi.ProgrammableInvoker: private native void initCifNativeThunkData(String[] argLayouts, String retLayout, boolean newArgTypes); */
+VM_BytecodeAction
+OutOfLineINL_jdk_internal_foreign_abi_ProgrammableInvoker_initCifNativeThunkData(J9VMThread *currentThread, J9Method *method)
+{
+	VM_BytecodeAction rc = EXECUTE_BYTECODE;
+	J9JavaVM *vm = currentThread->javaVM;
+	FFITypeHelpers FFIHelpers = FFITypeHelpers(currentThread);
+	ffi_cif *cif = NULL;
+	ffi_type *returnType = NULL;
+	ffi_type **argTypes = NULL;
+	J9CifArgumentTypes *cifArgTypesNode = NULL;
+
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	bool newArgTypes = (bool)(*(U_32*)currentThread->sp);
+	j9object_t retLayoutStrObject = J9_JNI_UNWRAP_REFERENCE(currentThread->sp + 1);
+	j9object_t argLayoutStrsObject = J9_JNI_UNWRAP_REFERENCE(currentThread->sp + 2);
+	j9object_t nativeInvoker = J9_JNI_UNWRAP_REFERENCE(currentThread->sp + 3);
+	U_32 argTypesCount = J9INDEXABLEOBJECT_SIZE(currentThread, argLayoutStrsObject);
+	returnType = (ffi_type *)FFIHelpers.getPrimitiveFFIType(retLayoutStrObject);
+
+	if (!newArgTypes) {
+		argTypes = (ffi_type **)(UDATA)J9VMJDKINTERNALFOREIGNABIPROGRAMMABLEINVOKER_ARGTYPESADDR(currentThread, nativeInvoker);
+	} else {
+		argTypes = (ffi_type **)j9mem_allocate_memory(sizeof(ffi_type *) * argTypesCount, OMRMEM_CATEGORY_VM);
+		if (NULL == argTypes) {
+			rc = GOTO_THROW_CURRENT_EXCEPTION;
+			setNativeOutOfMemoryError(currentThread, 0, 0);
+			goto done;
+		}
+
+		memset(argTypes, 0, sizeof(ffi_type *) * argTypesCount);
```

Comment:
Since you are writing all the elements in `argTypes` I dont think you need the memset

Reply:
Agreed and removed `memset()` as suggested.